rewrite-java/src/main/java/org/openrewrite/java/nullability/StandardizeNullabilityAnnotationsVisitor.java

```diff
@@ -0,0 +1,326 @@
+            NullabilityAnnotation nullabilityAnnotation = knownNullabilityDictionary.get(fullyQualified.getFullyQualifiedName());
+            getCursor().getParentOrThrow()
+                    .computeMessageIfAbsent(NULLABILITY_ANNOTATION_MARKER, k -> new HashSet<>())
+                    .add(new MatchedNullabilityAnnotation(jAnnotation, nullabilityAnnotation));
+        }
+        return jAnnotation;
+    }
+
+    @Override
+    public J.AnnotatedType visitAnnotatedType(J.AnnotatedType t, ExecutionContext executionContext) {
+        J.AnnotatedType annotatedType = super.visitAnnotatedType(t, executionContext);
+        Set<MatchedNullabilityAnnotation> matchedNullabilityAnnotations = getCursor().pollMessage(NULLABILITY_ANNOTATION_MARKER);
+        if (matchedNullabilityAnnotations == null || matchedNullabilityAnnotations.isEmpty()) {
+            return annotatedType;
+        }
+        Set<NullabilityAnnotation.Target> targetTypes = new HashSet<>();
+        targetTypes.add(NullabilityAnnotation.Target.TYPE_USE);
+        Cursor parentScopeCursor = getCursor().dropParentUntil(o -> o instanceof J.ClassDeclaration || o instanceof J.MethodDeclaration || Cursor.ROOT_VALUE.equals(o));
+        if (Cursor.ROOT_VALUE.equals(parentScopeCursor.getValue())) {
+            targetTypes.add(NullabilityAnnotation.Target.LOCAL_FIELD);
+        } else if (parentScopeCursor.getValue() instanceof J.MethodDeclaration) {
+            targetTypes.add(NullabilityAnnotation.Target.PARAMETER);
+        } else if (parentScopeCursor.getValue() instanceof J.ClassDeclaration) {
+            targetTypes.add(NullabilityAnnotation.Target.FIELD);
+        }
+        List<NullabilityAnnotation> annotationsForReplacement = getAnnotationsForReplacement(matchedNullabilityAnnotations, targetTypes);
+        if (annotationsForReplacement.isEmpty()) {
+            return annotatedType;
+        }
+
+        maybeRemoveMatchedAnnotationImports(matchedNullabilityAnnotations);
+
+        List<J.Annotation> currentNullabilityAnnotations = matchedNullabilityAnnotations.stream().map(MatchedNullabilityAnnotation::getJAnnotation).collect(Collectors.toList());
+        List<J.Annotation> cleanedAnnotations = new LinkedList<>(annotatedType.getAnnotations());
+        cleanedAnnotations.removeAll(currentNullabilityAnnotations);
+
+        J.AnnotatedType cleanAnnotatedType = annotatedType.withAnnotations(cleanedAnnotations);
+        J.AnnotatedType typeWithNewAnnotations = annotationsForReplacement.stream().reduce(
+                cleanAnnotatedType,
+                (J.AnnotatedType currentType, NullabilityAnnotation annotation) -> currentType.withTemplate(
+                        annotationTemplate(annotation),
+                        currentType.getCoordinates().addAnnotation(Comparator.comparing(J.Annotation::getSimpleName))
+                ), (oldType, newType) -> newType
+        );
```

Comment:
I assume you would like to insert the annotations at the same position where we removed them from. Basically we could do this using `List#indexOf` to retrieve the previous indices and later do the same inside the comparator for insertion.

We could remember the first index we removed an annotation from:
`int firstIndexWeRemoveAnnotations = jPackage.getAnnotations().indexOf(currentNullabilityAnnotations.get(0));`
Comparator respecting the remembered insertion point:
```java
(a1, a2) -> {
    int previousIndexA1 = jPackage.getAnnotations().indexOf(a1);
    int previousIndexA2 = jPackage.getAnnotations().indexOf(a2);
    if (previousIndexA1 < 0 && previousIndexA2 < 0) {
        return Comparator.comparing(J.Annotation::getSimpleName).compare(a1, a2);
    } else if (previousIndexA1 < 0) {
        return previousIndexA2 >= firstIndexWeRemoveAnnotations ? -1 : 1;
    } else if (previousIndexA2 < 0) {
        return previousIndexA1 < firstIndexWeRemoveAnnotations ? -1 : 1;
    } else {
        return 0;
    }
}
```
But this would always insert at the first position. In cases where we replace e. g. two annotations with two annotations would insert both new annotations at the position of the first that we removed.

I understand the reasoning behind this, but I am not sure how "perfect" we are able to achieve that. Also, the code snippet is quite long for such a "trivial purpose", but there's no api yet to insert at specific indices.

Reply:
Yes, it doesn't appear to be that easy at the moment, unfortunately. Let's keep it like this for now then. Thanks for clarifying!